tpl/tplimpl/templatestore_integration_test.go

```diff
@@ -684,6 +685,102 @@ layout: mylayout
 	b.AssertFileContent("public/en/foo/index.xml", "layouts/list.xml")
 }
 
+func TestLookupOrderIssue13636(t *testing.T) {
+	t.Parallel()
+
+	filesTemplate := `
+-- hugo.toml --
+defaultContentLanguage = "en"
+defaultContentLanguageInSubdir = true
+[languages]
+[languages.en]
+weight = 1
+[languages.nn]
+weight = 2
+-- content/s1/p1.en.md --
+---
+outputs: ["html", "amp", "json"]
+---
+-- content/s1/p1.nn.md --
+---
+outputs: ["html", "amp", "json"]
+---
+-- layouts/L1 --
+L1
+-- layouts/L2 --
+L2
+-- layouts/L3 --
+L3
+
+`
+
+	tests := []struct {
+		Lang       string
+		L1         string
+		L2         string
+		L3         string
+		ExpectHTML string
+		ExpectAmp  string
+		ExpectJSON string
+	}{
+		{"en", "all.en.html", "all.html", "single.html", "single.html", "single.html", ""},
+		{"en", "all.amp.html", "all.html", "page.html", "page.html", "all.amp.html", ""},
+		{"en", "all.amp.html", "all.html", "list.html", "all.html", "all.amp.html", ""},
+		{"en", "all.en.html", "all.json", "single.html", "single.html", "single.html", "all.json"},
+		{"en", "all.en.html", "single.json", "single.html", "single.html", "single.html", "single.json"},
+		{"en", "all.en.html", "all.html", "list.html", "all.en.html", "all.en.html", ""},
+		{"en", "list.en.html", "list.html", "list.en.html", "", "", ""},
+		{"nn", "all.en.html", "all.html", "single.html", "single.html", "single.html", ""},
+		{"nn", "all.en.html", "all.nn.html", "single.html", "single.html", "single.html", ""},
+		{"nn", "all.en.html", "all.nn.html", "single.nn.html", "single.nn.html", "single.nn.html", ""},
+		{"nn", "single.json", "single.nn.json", "all.json", "", "", "single.nn.json"},
+		{"nn", "single.json", "single.en.json", "all.nn.json", "", "", "single.json"},
```

Comment:
@jmooring OK, so I finally understood what the issue was about, and you were right, there were some issues to be fixed. I'm not sure if the above table matches your table, but it's stable. The only question I had after this was the line above:

>{"nn", "single.json", "single.en.json", "all.nn.json", "", "", "single.json"},

For lang `nn` we now prefer `single.json` over `all.nn.json`, which in my head makes the most sense, but I'm not entirely sure.

Anyway, thanks for the bug report (I thought we had decent test coverage on this ....), and I would appreciate if you could take this branch for a spin.